internal/sorter/sorter.go

```diff
@@ -296,30 +296,36 @@ func (s routeSorter) Less(i, j int) bool {
 	switch a := s[i].PathMatchCondition.(type) {
 	case *dag.PrefixMatchCondition:
 		if b, ok := s[j].PathMatchCondition.(*dag.PrefixMatchCondition); ok {
-			cmp := strings.Compare(a.Prefix, b.Prefix)
-			switch cmp {
-			case 1:
+			switch {
+			case len(a.Prefix) > len(b.Prefix):
 				// Sort longest prefix first.
 				return true
-			case -1:
+			case len(a.Prefix) < len(b.Prefix):
 				return false
 			default:
-				if a.PrefixMatchType == b.PrefixMatchType {
-					return compareRoutesByMethodHeaderQueryParams(s[i], s[j])
+				cmp := strings.Compare(a.Prefix, b.Prefix)
+				switch cmp {
+				case 1:
+					return true
+				case 2:
+					return false
+				default:
+					if a.PrefixMatchType == b.PrefixMatchType {
+						return compareRoutesByMethodHeaderQueryParams(s[i], s[j])
+					}
+					// Segment prefixes sort first as they are more specific.
+					return a.PrefixMatchType == dag.PrefixMatchSegment
 				}
-				// Segment prefixes sort first as they are more specific.
-				return a.PrefixMatchType == dag.PrefixMatchSegment
 			}
 		}
 	case *dag.RegexMatchCondition:
 		switch b := s[j].PathMatchCondition.(type) {
 		case *dag.RegexMatchCondition:
-			cmp := strings.Compare(a.Regex, b.Regex)
-			switch cmp {
-			case 1:
+			switch {
+			case len(a.Regex) > len(b.Regex):
 				// Sort longest regex first.
 				return true
-			case -1:
+			case len(a.Regex) < len(b.Regex):
 				return false
 			default:
```

Comment:
since we store routes and access them as a map, i think we will need some sort of fallback to sort them stably if the lengths are the same, otherwise we may have routes change order

Reply:
nit: can we add a test case for this scenario?

Reply:
I believe a test already exists in `TestSortRoutesLongestHeaders` where everything is `PathMatchCondition: matchRegex("/pathregex"),`

Reply:
see this test case:

```diff
diff --git a/internal/sorter/sorter_test.go b/internal/sorter/sorter_test.go
index 90a158c8..98256a84 100644
--- a/internal/sorter/sorter_test.go
+++ b/internal/sorter/sorter_test.go
@@ -284,6 +284,12 @@ func TestSortRoutesPathMatch(t *testing.T) {
                {
                        PathMatchCondition: matchRegex(`/foo((\/).*)*`),
                },
+               {
+                       PathMatchCondition: matchRegex("/bar.*"),
+               },
+               {
+                       PathMatchCondition: matchRegex("/foo.*"),
+               },
                {
                        PathMatchCondition: matchRegex("/"),
                },
@@ -712,6 +718,6 @@ func shuffleAndCheckSort[T any](t *testing.T, want []T) {
                have := shuffleSlice(want)

                sort.Stable(For(have))
-               assert.Equal(t, want, have)
+               assert.Equalf(t, want, have, "failed during trial %d", i)
        }
 }
```

we run all these sorter tests through a bunch of trials and you'll see if you run `TestSortRoutesPathMatch` that the test fails a few times but not all with either ordering of those two route matches, since those regexes are the same length

can be fixed if we have a fallback to sorting lexicographically like we do with prefix matches (and having the correct expected order)

Reply:
I think I stabilized it now with de92073f94f2426279a4f99d39029d97355b61d2